azavea/tilegarden
# Feature/s3 config

## Overview

This PR allows you to load configuration files from an S3 bucket rather than only locally. Adding a query string such as `?config=config-file.xml&s3bucket=tilegarden-bucket` will get you the config file with the S3 key "config-file.xml" that is stored in the bucket "tilegarden-bucket", as long as you deployed the lambda function using AWS credentials that have S3 read/write access. It will also work in local development, provided such AWS credentials are specified in your `.env`. 

In pursuit of this, a new script `build-xml` has been added to build a directory's worth of MML and MSS into a directory of XML without having to spin up the entire development apparatus and dig for the compiled files in `tiler/bin/`. It has several options that look like the following:
```
Usage: build-xml -i <path> -o <path>
Transpiles a directory's worth of MML+MSS files into Tilegarden-readable XML
Options:
    -h, --help          Display this help text
    -i, --in <path>     Relative path to input directory [src/tiler/src/config]
    -o, --out <path>    Relative path to output directory [transpiled-configs/]
    -b, --bucket <name> Uploads XML to the specified S3 bucket, instead of saving to disk (requires AWS CLI installed on your machine, uses credentials from .env)
    -d, --development   Build using development variables, instead of production ones
```
This script lead to a re-write of the whole XML-building pipeline. Most of the logic was relegated to bash scripts, and Carto was removed as a project dependency and instead installed as a global utility on the tiler container. 

### Demo
```
$ ./scripts/build-xml -d -b tilegarden-dev
Transpiling src/tiler/src/config//data-type-example.mml => /tmp/tmp.naNUwi4M32/data-type-example.xml
Starting tilegarden_database_1 ... done
Transpiling src/tiler/src/config//map-config.mml => /tmp/tmp.naNUwi4M32/map-config.xml
Starting tilegarden_database_1 ... done
Transpiling src/tiler/src/config//point-example.mml => /tmp/tmp.naNUwi4M32/point-example.xml
Starting tilegarden_database_1 ... done
Uploading files to s3://tilegarden-dev/...
upload: ../../../../tmp/tmp.naNUwi4M32/data-type-example.xml to s3://tilegarden-dev/data-type-example.xml
upload: ../../../../tmp/tmp.naNUwi4M32/map-config.xml to s3://tilegarden-dev/map-config.xml
upload: ../../../../tmp/tmp.naNUwi4M32/point-example.xml to s3://tilegarden-dev/point-example.xml
Cleaning up /tmp/tmp.naNUwi4M32...
```

`http://localhost:3000/tile/13/2385/3103.png?config=point-example.xml&s3bucket=tilegarden-dev`:
![image](https://user-images.githubusercontent.com/16990679/43611820-663f9896-9678-11e8-86cc-e84052fcd98b.png)

## Testing Instructions
 * Build the docker containers with `./scripts/update`.
 * Use `build-xml` to create some XML files and upload them to the S3 bucket of your choice. 
   * _Ex._ `./scripts/build-xml --development --bucket tilegarden-dev` to build XML files with development database connection info from the default input directory at `src/tiller/src/config` and push them to the S3 bucket "tilegarden-dev" that I've already created.
   * If you use the "tilegarden-dev" bucket, there might already be config files in there, so it might be a good idea to modify/rename the MSS/MML file you're testing with so you can note the difference.
 * Start the development server with `./scripts/server`.
 * Navigate to `http://localhost:3000/tile/13/2385/3103.png?config=<config file name>&s3bucket=<S3 bucket name>`. Switch between different config files to see what happens.

Resolves #26
